### DISCUSSION TITLE
Error / Failure Messages On Rollups

### DISCUSSION THREAD

**Discussion by TanukiDemon** [jamessimone/apex-rollup#284] [Q&A] [answered]
Hi @jamessimone! First off, very cool project you've got rolling here. I've been working with it in a sandbox and love how simple and intuitive most things are. In particular I like the Recalculate Rollup lightning tab, and the ability to run the rollups off of CMDT. It's there, however, that I've run into a few issues. I have 4 Count rollups defined for the Opportunity object, and 2 of them can be recalculated with ease, but two of them fail with just this message: "Rollup job status: Failed".

A separate issue I've run into was in attempting to roll up a custom object, which resulted in the following error message: "There was an error performing your rollup: Ending position out of bounds: -1". Any tips on how I can troubleshoot these? I'll try to dig a little deeper on my own in the meantime.

**Comment by jamessimone** [comment, marked as answer]
Hi @TanukiDemon and thanks for creating this discussion. Your where clauses are likely leading to the errors above — you have to have spaces between the fields, the comparison (=, <>, !=, etc) and the values you’re comparing to. Syntactically, this is not valid for Apex Rollup:

`Account.Name=‘foo’`

The correct statement thus becomes:

`Account.Name = ‘foo’`

Hopefully that helps!

**Reply by TanukiDemon** [reply]
Thanks for the incredibly speedy response @jamessimone. That seems to have been spot on for the ending position out of bounds errors, though I seem to still have issues with the simple "Failure" message. Here's the where clause for that Opportunity Rollup

RecordTypeId = '01236000000xgbdAAA' AND Event_Date__c <= LAST_YEAR AND Event_Date__c > N_YEARS_AGO:3 AND StageName != 'Closed Lost' AND StageName != 'Closed/No Event Fees Only'

Edited as I originally included the where clause for another rollup that had the same issue, but didn't match the screenshot

I'm also including an image of the entire rollup record. Does anything jump out at you? 

![OppRollup](https://user-images.githubusercontent.com/11148381/163248558-129708f0-99df-4d40-9e4b-d257bc8c9a06.png)

**Reply by jamessimone** [reply]
@TanukiDemon there's nothing immediately wrong with that where clause that I can see, but I may need to see a log in order to identify what the issue - I would also encourage you to take a look through the logs that get generated when you use the Recalculate Rollup app when you have `Is Rollup Logging Enabled` on the Org Default Rollup Control record - it's possible that the messages printed there will help to address whatever the issue may be that you're having.

**Reply by TanukiDemon** [reply]
Thanks James. I'll look into it and report back if I find anything.

**Reply by TanukiDemon** [reply]
@jamessimone Thanks for the tip on how to do the rollup logging; looks like the N_YEARS_AGO relative date value wasn't recognized as a date. I bolded the relevant line in the debug log below!


07:18:21.0 (753125652)|SYSTEM_METHOD_ENTRY|[729]|String.valueOf(Object)
07:18:21.0 (753137195)|HEAP_ALLOCATE|[729]|Bytes:13
07:18:21.0 (753148663)|SYSTEM_METHOD_EXIT|[729]|String.valueOf(Object)
07:18:21.0 (753155496)|VARIABLE_SCOPE_BEGIN|[729]|comparisonText|String|false|false
07:18:21.0 (753169668)|VARIABLE_ASSIGNMENT|[729]|comparisonText|"N_YEARS_AGO:3"
07:18:21.0 (753189080)|STATEMENT_EXECUTE|[731]
07:18:21.0 (753191241)|STATEMENT_EXECUTE|[732]
07:18:21.0 (753207028)|SYSTEM_METHOD_ENTRY|[732]|com.salesforce.api.interop.apex.bcl.DateMethods.valueOf(String)
07:18:21.0 (753370283)|HEAP_ALLOCATE|[732]|Bytes:31
07:18:21.0 (753386461)|SYSTEM_METHOD_EXIT|[732]|com.salesforce.api.interop.apex.bcl.DateMethods.valueOf(String)
07:18:21.0 (753399295)|METHOD_EXIT|[649]|01pK0000001hTHd|RollupEvaluator.WhereFieldCondition.getGreaterOrLessThan(Object, Object)
07:18:21.0 (753408055)|METHOD_EXIT|[542]|01pK0000001hTHd|RollupEvaluator.WhereFieldCondition.equals(Object)
07:18:21.0 (753417210)|METHOD_EXIT|[500]|01pK0000001hTHd|RollupEvaluator.AndConditionalGrouping.conditionallyEquals(Object)
07:18:21.0 (753425081)|METHOD_EXIT|[251]|01pK0000001hTHd|RollupEvaluator.ConditionalGrouping.equals(Object)
07:18:21.0 (753438813)|METHOD_EXIT|[243]|01pK0000001hTHd|RollupEvaluator.WhereFieldEvaluator.innerMatches(Object)
07:18:21.0 (753449264)|METHOD_EXIT|[2733]|01pK0000001hTHd|RollupEvaluator.WhereFieldEvaluator.matches(Object)
07:18:21.0 (753464831)|METHOD_EXIT|[2436]|01pK0000001hTHV|Rollup.filter(List<SObject>, Map<Id,SObject>, Rollup.Evaluator, Rollup__mdt, Schema.SObjectType, RollupCalcItemReplacer)
07:18:21.0 (753480005)|METHOD_EXIT|[292]|01pK0000001hTHV|Rollup.getRollup(List<Rollup__mdt>, Schema.SObjectType, List<SObject>, Map<Id,SObject>, Rollup.Evaluator, Rollup.InvocationPoint)
07:18:21.0 (753491538)|METHOD_EXIT|[480]|01pK0000001hTHV|Rollup.getAsyncRollup(List<Rollup__mdt>, Schema.SObjectType, List<SObject>, Map<Id,SObject>, Rollup.Evaluator, Rollup.InvocationPoint)
07:18:21.0 (753503411)|METHOD_EXIT|[27]|01pK0000001hTHW|RollupAsyncProcessor.getDelegatedFullRecalcRollup(List<Rollup__mdt>, List<SObject>, RollupAsyncProcessor.FullRecalcProcessor)
07:18:21.0 (753514773)|METHOD_EXIT|[17]|01pK0000001hTHc|RollupDeferredFullRecalcProcessor.getProcessor()
07:18:21.0 (753524718)|METHOD_EXIT|[652]|01pK0000001hTHc|RollupDeferredFullRecalcProcessor.transformFullRecalcRollups()
07:18:21.0 (753534119)|METHOD_EXIT|[567]|01pK0000001hTHW|RollupAsyncProcessor.transformRollups(List<RollupAsyncProcessor>)
07:18:21.0 (753546121)|SYSTEM_MODE_EXIT|true
07:18:21.0 (753557661)|METHOD_EXIT|[450]|01pK0000001hTHW|RollupAsyncProcessor.process(List<RollupAsyncProcessor>)
07:18:21.0 (753565126)|SYSTEM_MODE_EXIT|false
**07:18:21.0 (753656877)|FATAL_ERROR|System.TypeException: Invalid date: N_YEARS_AGO:3**

Class.RollupEvaluator.WhereFieldCondition.getGreaterOrLessThan: line 732, column 1
Class.RollupEvaluator.WhereFieldCondition.equals: line 649, column 1
Class.RollupEvaluator.AndConditionalGrouping.conditionallyEquals: line 542, column 1
Class.RollupEvaluator.ConditionalGrouping.equals: line 500, column 1
Class.RollupEvaluator.WhereFieldEvaluator.innerMatches: line 251, column 1
Class.RollupEvaluator.WhereFieldEvaluator.matches: line 243, column 1
Class.Rollup.filter: line 2733, column 1
Class.Rollup.getRollup: line 2436, column 1
Class.Rollup.getAsyncRollup: line 292, column 1
Class.RollupAsyncProcessor.getDelegatedFullRecalcRollup: line 480, column 1
Class.RollupDeferredFullRecalcProcessor.getProcessor: line 27, column 1
Class.RollupDeferredFullRecalcProcessor.transformFullRecalcRollups: line 17, column 1
Class.RollupAsyncProcessor.transformRollups: line 652, column 1
Class.RollupAsyncProcessor.process: line 567, column 1
Class.RollupAsyncProcessor.QueueableProcessor.execute: line 450, column 1
07:18:21.0 (754116238)|CODE_UNIT_FINISHED|RollupDeferredFullRecalcProcessor
07:18:21.0 (754127256)|EXECUTION_FINISHED

**Reply by jamessimone** [reply]
@TanukiDemon haha we arrived at the same conclusion by different means - I started testing this about an hour ago as well! I just sent @ssk42 a message - we worked on the Date Literal support together - but it looks like the [Salesforce Developer Docs page for Date Literals](https://developer.salesforce.com/docs/atlas.en-us.198.0.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_select_dateformats.htm) is missing the following:

- N_DAYS_AGO
- N_WEEKS_AGO
- N_MONTHS_AGO
- N_QUARTERS_AGO
- N_FISCAL_QUARTERS_AGO
- N_YEARS_AGO
- N_FISCAL_YEARS_AGO

I will work to get this addressed within Apex Rollup - in the meantime, you should be good to switch to using `LAST_N_YEARS` for now. Let me know!
